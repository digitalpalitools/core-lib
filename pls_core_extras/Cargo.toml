[package]
name = "pls_core_extras"
version = "0.1.0"
authors = ["Digital Pāli Tools <digitalpalitools@gmail.com>"]
edition = "2018"
description = "Extra supporting implementations for pls_core meanign for non-web usage."

[dependencies]
lazy_static = "1.4.0"
pls_core = { path = '../pls_core' }
serde_json = "1.0.63"
colored = "2"
chrono = "0.4"

[dependencies.rusqlite]
version = "0.24.2"
features = ["bundled"]

[dev-dependencies]
proptest = "0.10.1"
test-case = "1.1.0"
insta = "1.7.0"
