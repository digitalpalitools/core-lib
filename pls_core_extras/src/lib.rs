#[macro_use]
extern crate lazy_static;

pub mod host;
pub mod inflection_generator;
pub mod logger;
pub mod sql_access;
