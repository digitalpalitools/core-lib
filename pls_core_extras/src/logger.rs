use chrono::Local;
use colored::*;

fn get_time_stamp() -> String {
    Local::now().format("%y-%m-%d %H:%M:%S").to_string()
}

pub trait PlsLogger {
    fn info(&self, msg: &str);
    fn error(&self, msg: &str);
    fn warning(&self, msg: &str);
}

pub struct ColoredConsoleLogger;

impl PlsLogger for ColoredConsoleLogger {
    fn info(&self, msg: &str) {
        println!(
            "{} {}",
            get_time_stamp().white(),
            format!("info: {}", msg).green(),
        );
    }

    fn error(&self, msg: &str) {
        println!(
            "{} {}",
            get_time_stamp().white(),
            format!("error: {}", msg).red(),
        );
    }

    fn warning(&self, msg: &str) {
        println!(
            "{} {}",
            get_time_stamp().white(),
            format!("warning: {}", msg).yellow(),
        );
    }
}
