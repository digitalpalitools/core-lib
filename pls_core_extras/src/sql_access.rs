use rusqlite::{Connection, Row, NO_PARAMS};

pub struct SqlAccess {
    pub connection: Connection,
}

impl SqlAccess {
    fn get_row_cells(&self, row: &Row) -> Vec<String> {
        let cells: Vec<String> = row
            .column_names()
            .iter()
            .map(|&cn| {
                let cell: String = match row.get(cn) {
                    Ok(c) => c,
                    Err(e) => e.to_string(),
                };
                cell
            })
            .collect();

        cells
    }

    pub fn exec_sql_core(&self, sql: &str) -> rusqlite::Result<Vec<Vec<Vec<String>>>, String> {
        let mut result: Vec<Vec<Vec<String>>> = Vec::new();
        for s in sql.split(';').filter(|s| !s.trim().is_empty()) {
            let mut stmt = self.connection.prepare(&s).map_err(|e| e.to_string())?;
            let mut rows = stmt.query(NO_PARAMS).map_err(|e| e.to_string())?;

            let mut table: Vec<Vec<String>> = Vec::new();
            while let Some(row) = rows.next().map_err(|e| e.to_string())? {
                table.push(self.get_row_cells(row));
            }
            result.push(table)
        }

        Ok(result)
    }
}
