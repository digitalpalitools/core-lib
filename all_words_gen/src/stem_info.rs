use pls_core_extras::sql_access::SqlAccess;
use std::fmt;

#[derive(Debug)]
pub struct StemInfo {
    pub row_id: i64,
    pub pali1: String,
}

pub struct StemInfoIterator<'a> {
    sql_access: &'a SqlAccess,
    max_stems_to_fetch: i64,
    max_batch_size: i64,
    pub stems_fetched: i64,
    pub batches_fetched: i64,
    pub error: Option<String>,
}

impl fmt::Display for StemInfoIterator<'_> {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        writeln!(f, "Iterator state:")?;
        writeln!(f, "... max_stems_to_fetch: {}", self.max_stems_to_fetch)?;
        writeln!(f, "... max_batch_size: {}", self.max_batch_size)?;
        writeln!(f, "... stems_fetched: {}", self.stems_fetched)?;
        writeln!(f, "... batches_fetched: {}", self.batches_fetched)?;
        writeln!(f, "... error: {:?}", self.error)?;
        Ok(())
    }
}

impl<'a> StemInfoIterator<'a> {
    pub fn new(sql_access: &'a SqlAccess, max_stems_to_fetch: i64, max_batch_size: i64) -> Self {
        StemInfoIterator {
            sql_access,
            max_stems_to_fetch,
            max_batch_size,
            stems_fetched: 0,
            batches_fetched: 0,
            error: None,
        }
    }
}

impl Iterator for StemInfoIterator<'_> {
    type Item = Vec<StemInfo>;

    fn next(&mut self) -> Option<Self::Item> {
        let stems_remaining = self.max_stems_to_fetch - self.stems_fetched;
        let batch_size = if stems_remaining < self.max_batch_size {
            stems_remaining
        } else {
            self.max_batch_size
        };

        let curr_batch = self
            .sql_access
            .exec(&format!(
                "SELECT CAST(rowid AS TEXT), pāli1 FROM '_stems' order by rowid LIMIT {},{}",
                self.stems_fetched, batch_size
            ))
            .unwrap_or_else(|x| {
                self.error = Some(x);
                vec![]
            });

        if curr_batch.len() == 1
            && !curr_batch[0].is_empty()
            && (self.stems_fetched < self.max_stems_to_fetch)
        {
            self.batches_fetched += 1;
            let stem_infos: Vec<StemInfo> = curr_batch[0]
                .to_owned()
                .into_iter()
                .map(|x| StemInfo {
                    row_id: x[0]
                        .parse::<i64>()
                        .expect("rowid must be i64. check for db corruption."),
                    pali1: x[1].to_owned(),
                })
                .collect();
            self.stems_fetched += stem_infos.len() as i64;

            Some(stem_infos)
        } else {
            None
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use pls_core_extras::sql_access::create_sql_access;

    #[test]
    fn test_iterator_uneven_batches() {
        let sa = create_sql_access();
        let mut sii = StemInfoIterator::new(&sa, 1000, 11);
        for _si in &mut sii {}

        assert_eq!(sii.max_stems_to_fetch, 1000);
        assert_eq!(sii.max_batch_size, 11);
        assert_eq!(sii.stems_fetched, 1000);
        assert_eq!(sii.batches_fetched, 91);
        assert_eq!(sii.error, None);
    }

    #[test]
    fn test_iterator_even_batches() {
        let sa = create_sql_access();
        let mut sii = StemInfoIterator::new(&sa, 1000, 10);
        for _si in &mut sii {}

        assert_eq!(sii.max_stems_to_fetch, 1000);
        assert_eq!(sii.max_batch_size, 10);
        assert_eq!(sii.stems_fetched, 1000);
        assert_eq!(sii.batches_fetched, 100);
        assert_eq!(sii.error, None);
    }

    #[test]
    fn test_iterator_trailing_batches() {
        let sa = create_sql_access();
        let mut sii = StemInfoIterator::new(&sa, 25, 10);
        for _si in &mut sii {}

        assert_eq!(sii.max_stems_to_fetch, 25);
        assert_eq!(sii.max_batch_size, 10);
        assert_eq!(sii.stems_fetched, 25);
        assert_eq!(sii.batches_fetched, 3);
        assert_eq!(sii.error, None);
    }
}
