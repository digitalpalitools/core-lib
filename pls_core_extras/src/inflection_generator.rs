use crate::host::PlsHost;
use crate::logger::PlsLogger;
use crate::sql_access::SqlAccess;
use pls_core::inflections::{
    generate_all_inflections, generate_inflection_table, host::PlsInflectionsHost,
};
use rusqlite::Connection;
use std::env;

lazy_static! {
    static ref PLS_INFLECTION_GENERATOR_PREFIX: String =
        env::var("__PLS_INFLECTION_GENERATOR_PREFIX__").unwrap_or_else(|_e| "".to_string());
}

pub trait InflectionGenerator {
    fn check_inflection_db(&self, logger: &dyn PlsLogger) -> Result<(), String>;
    fn generate_inflection_table_html(&self, pali1: &str) -> String;
    fn generate_all_inflections(&self, pali1: &str) -> Vec<String>;
}

pub struct NullInflectionGenerator {}

impl NullInflectionGenerator {
    pub fn new() -> NullInflectionGenerator {
        NullInflectionGenerator {}
    }
}

impl Default for NullInflectionGenerator {
    fn default() -> Self {
        Self::new()
    }
}

impl InflectionGenerator for NullInflectionGenerator {
    fn check_inflection_db(&self, _logger: &dyn PlsLogger) -> Result<(), String> {
        Ok(())
    }

    fn generate_inflection_table_html(&self, _pali1: &str) -> String {
        "".to_string()
    }

    fn generate_all_inflections(&self, _pali1: &str) -> Vec<String> {
        Vec::new()
    }
}

pub struct PlsInflectionGenerator<'a> {
    inflection_host: PlsHost<'a>,
}

impl<'a> PlsInflectionGenerator<'a> {
    pub fn new(
        locale: &'a str,
        version: &'a str,
        url: &'a str,
        inflections_db_path: &'a str,
        logger: &'a dyn PlsLogger,
    ) -> Result<PlsInflectionGenerator<'a>, String> {
        logger.info(&format!(
            "Open inflections db: '{}'...",
            inflections_db_path
        ));
        let connection = Connection::open(inflections_db_path)
            .map_err(|e| format!("Cannot open db '{}'. Error: {}.", inflections_db_path, e))?;

        let igen = PlsInflectionGenerator {
            inflection_host: PlsHost {
                locale,
                version,
                url,
                sql_access: SqlAccess { connection },
                logger,
            },
        };

        Ok(igen)
    }
}

impl<'a> InflectionGenerator for PlsInflectionGenerator<'a> {
    fn check_inflection_db(&self, logger: &dyn PlsLogger) -> Result<(), String> {
        match self
            .inflection_host
            .exec_sql_query("select * from _version")
        {
            Ok(ver_info) => {
                if ver_info.len() != 1 && ver_info[0].len() != 1 && ver_info[0][0].len() != 3 {
                    let msg =
                        "Invalid inflection db. Unexpected data in _version table.".to_string();
                    logger.error(&msg);
                    return Err(msg);
                }

                let commit_id = &ver_info[0][0][0];
                let repository = &ver_info[0][0][2];
                if commit_id.len() != 40 && repository != "digitalpalitools/inflection-generator" {
                    let msg = format!(
                        "Invalid inflection db: commid_id: {}, repository: {}.",
                        commit_id, repository
                    );
                    logger.error(&msg);
                    return Err(msg);
                }

                logger.info(&format!(
                    "... loaded version: https://github.com/{}#{}.",
                    repository,
                    &commit_id[0..10]
                ));
                Ok(())
            }
            Err(e) => {
                let msg = format!("Unable to load inflection db. Error: {}.", e);
                logger.error(&msg);
                Err(e)
            }
        }
    }

    fn generate_inflection_table_html(&self, pali1: &str) -> String {
        if is_black_listed_word(pali1) {
            return "".to_string();
        }

        match generate_inflection_table(pali1, false, &self.inflection_host) {
            Ok(t) => t,
            Err(e) => {
                self.inflection_host.logger.warning(&format!(
                    "Unable to generate inflection table '{}'. Error: {}.",
                    pali1, e
                ));
                "".to_string()
            }
        }
    }

    fn generate_all_inflections(&self, pali1: &str) -> Vec<String> {
        if is_black_listed_word(pali1) {
            return vec![];
        }

        match generate_all_inflections(pali1, &self.inflection_host) {
            Ok(inflections) => inflections,
            Err(e) => {
                self.inflection_host.logger.warning(&format!(
                    "Unable to generate inflections for '{}'. Error: {}.",
                    pali1, e
                ));
                Vec::new()
            }
        }
    }
}

fn is_black_listed_word(pali1: &str) -> bool {
    let prefix: &str = &PLS_INFLECTION_GENERATOR_PREFIX;
    !prefix.is_empty() && !pali1.starts_with(prefix)
}
