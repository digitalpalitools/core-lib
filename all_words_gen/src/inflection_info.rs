use crate::stem_info::StemInfo;
use pls_core_extras::inflection_generator::InflectionGenerator;
use serde::Serialize;

#[derive(Serialize, Debug)]
pub struct InflectionInfo {
    pub pali1_id: i64,
    pub inflection: String,
}

#[derive(Serialize, Debug)]
pub struct InflectionsBatchInfo {
    pub inflection_infos: Vec<InflectionInfo>,
    pub inflected_forms_fetched: i64,
}

pub fn create_inflection_infos(
    stem_infos: Vec<StemInfo>,
    igen: &dyn InflectionGenerator,
) -> InflectionsBatchInfo {
    let mut ibi = InflectionsBatchInfo {
        inflection_infos: vec![],
        inflected_forms_fetched: 0,
    };

    for stem_info in stem_infos {
        let mut infs = igen.generate_all_inflections(&stem_info.pali1);
        infs.sort();
        infs.dedup();
        if infs.is_empty() {
            ibi.inflected_forms_fetched += 1;
        }

        let mut inf_infos: Vec<InflectionInfo> = infs
            .into_iter()
            .map(|inf| InflectionInfo {
                pali1_id: stem_info.row_id,
                inflection: inf,
            })
            .collect();
        ibi.inflection_infos.append(&mut inf_infos);
    }

    ibi
}

#[cfg(test)]
mod tests {
    use super::*;
    use pls_core_extras::inflection_generator::PlsInflectionGenerator;
    use pls_core_extras::logger::{NullLogger, PlsLogger};
    use pls_core_extras::sql_access::resolve_file_in_manifest_dir;

    #[test]
    fn test_create_inflection_infos() {
        let db_path = &get_db_path();
        let igen = create_inflection_generator(db_path, &NullLogger {});
        let stems = vec![
            StemInfo {
                row_id: 7,
                pali1: "ababa 1".to_string(),
            },
            StemInfo {
                row_id: 8,
                pali1: "ababa 2".to_string(),
            },
            StemInfo {
                row_id: 32,
                pali1: "abbahe".to_string(),
            },
        ];
        let infl_infos = create_inflection_infos(stems, &igen);

        insta::assert_yaml_snapshot!(infl_infos);
    }

    fn create_inflection_generator<'a>(
        db_path: &'a str,
        logger: &'a dyn PlsLogger,
    ) -> PlsInflectionGenerator<'a> {
        let igen = PlsInflectionGenerator::new(
            "en",
            env!("CARGO_PKG_VERSION"),
            env!("CARGO_PKG_NAME"),
            db_path,
            logger,
        )
        .expect("unexpected test setup failure");

        igen
    }

    fn get_db_path() -> String {
        resolve_file_in_manifest_dir("inflections.db")
            .expect("must exist")
            .as_path()
            .to_str()
            .expect("must exist")
            .to_string()
    }
}
