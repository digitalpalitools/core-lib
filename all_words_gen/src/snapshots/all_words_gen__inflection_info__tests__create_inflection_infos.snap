---
source: all_words_gen/src/inflection_info.rs
expression: infl_infos

---
inflection_infos:
  - pali1_id: 7
    inflection: ababa
  - pali1_id: 7
    inflection: ababamhi
  - pali1_id: 7
    inflection: ababamhā
  - pali1_id: 7
    inflection: ababasmiṃ
  - pali1_id: 7
    inflection: ababasmā
  - pali1_id: 7
    inflection: ababassa
  - pali1_id: 7
    inflection: ababato
  - pali1_id: 7
    inflection: ababaṃ
  - pali1_id: 7
    inflection: ababe
  - pali1_id: 7
    inflection: ababebhi
  - pali1_id: 7
    inflection: ababehi
  - pali1_id: 7
    inflection: ababena
  - pali1_id: 7
    inflection: ababesu
  - pali1_id: 7
    inflection: ababā
  - pali1_id: 7
    inflection: ababānaṃ
  - pali1_id: 7
    inflection: ababāni
  - pali1_id: 7
    inflection: ababāya
  - pali1_id: 8
    inflection: ababa
  - pali1_id: 8
    inflection: ababamhi
  - pali1_id: 8
    inflection: ababamhā
  - pali1_id: 8
    inflection: ababasmiṃ
  - pali1_id: 8
    inflection: ababasmā
  - pali1_id: 8
    inflection: ababassa
  - pali1_id: 8
    inflection: ababato
  - pali1_id: 8
    inflection: ababaṃ
  - pali1_id: 8
    inflection: ababe
  - pali1_id: 8
    inflection: ababebhi
  - pali1_id: 8
    inflection: ababehi
  - pali1_id: 8
    inflection: ababena
  - pali1_id: 8
    inflection: ababesu
  - pali1_id: 8
    inflection: ababā
  - pali1_id: 8
    inflection: ababānaṃ
  - pali1_id: 8
    inflection: ababāni
  - pali1_id: 8
    inflection: ababāya
inflected_forms_fetched: 1

